src/pyEQL/solution.py

```diff
@@ -130,6 +211,68 @@ def __init__(
             Temperature: 293.150 K
             Components: ['H2O(aq)', 'H[+1]', 'OH[-1]', 'Na[+1]', 'Cl[-1]']
         """
+        # Initialize basic solution attributes
+        self.volume = ureg.Quantity(volume or "1 L")
+        self.temperature = ureg.Quantity(temperature)
+        self.pressure = ureg.Quantity(pressure)
+        self.pH = pH
+        self.pE = pE
+        self.balance_charge = balance_charge
+        self.solvent = solvent
+        self.engine = engine
+        self.database = database
+        self.default_diffusion_coeff = default_diffusion_coeff
+        self.logger = logging.getLogger("pyEQL.Solution")
+        if log_level:
+            self.logger.setLevel(log_level)
+
+        # Initialize solution components
+        self.components = {}
+
+        # Parse solutes
+        if solutes:
+            if isinstance(solutes, dict):
+                for formula, amount in solutes.items():
+                    self.add_solute(formula, amount)
+            elif isinstance(solutes, list) and all(isinstance(i, list) and len(i) == 2 for i in solutes):
+                for formula, amount in solutes:
+                    self.add_solute(formula, amount)
+            else:
+                raise ValueError("Solutes must be a dict or a list of [formula, amount] pairs.")
+
+        # Ensure volume consistency
+        self.volume_update_required = False
+
+        # Handle special units like "%" and "ppb"
+        for formula, amount in self.components.items():
+            if isinstance(amount, str) and ("%" in amount or "ppb" in amount):
+                try:
+                    self.components[formula] = ureg.Quantity(amount)
+                except UndefinedUnitError:
+                    self.logger.error(f"Unit {amount} for {formula} is not defined in the unit registry.")
+
+        def preprocess_units(amount: str) -> str:
+        """
+        Convert unsupported units into pint-compatible equivalents.
+        
+        Args:
+            amount: The input quantity string, e.g., '1000 ppb'.
+
+        Returns:
+            A pint-compatible quantity string.
+        """
+        if "ppb" in amount:
+            value = amount.replace("ppb", "").strip()
+            return f"{value} microgram/L"
+        elif "ppm" in amount:
+            value = amount.replace("ppm", "").strip()
+            return f"{value} milligram/L"
+        elif "%" in amount:
+            value = amount.replace("%", "").strip()
+            return f"{value} dimensionless"
+        return amount  # Return unchanged if no preprocessing needed
+
+        
```

Comment:
There is an indentation error here - you have defines a new method in the middle of the `__init__` method. I also see that you have a similar, but not exactly the same `preprocess_units` defined at the top of the file.

You need to merge the two methods togethers, removing the one at the top of the file and then  moving this one so it comes after `__init__`.